New WIP resposive image patch for core

joemcgill opened the pull request:
Upstream changes got mixed in with the old PR and GitHub wasn't updating the diffs. This is cleaner.

joemcgill committed:
Add tests for srcset functions

* Adds new test image, image-large.png
* Adds new tests for `wp_attachment_img_srcset_array()`

joemcgill committed:
Add base responsive image functionality

* Extend `wp_get_attachment_image` with `srcset`/`sizes` attributes
* Add `wp_attachment_img_srcset()` and `wp_attachment_img_srcset_array()`

joemcgill committed:
Add `create_upload_object`  method to Factory->attachment

joemcgill committed:
Simplify tests using  method

joemcgill committed:
Add `srcset` and `sizes` to `get_image_tag()`

joemcgill committed:
Fix failing tests after 1a26728

joemcgill committed:
Test: update inline doc in `test_wp_attachment_img_srcset_array_with_edits`

joemcgill committed:
Tests: clean up custom image sizes from media tests

joemcgill committed:
Add sanity checks for meta values in `wp_attachment_img_srcset_array`

This fixes errors in existing media tests that create attachments for galleries without
producing all of the normal meta values. While this is an edge case, we should test
for the existence of meta keys rather than assuming they'll be there.

joemcgill committed:
Escape srcset and sizes attributes on output

joemcgill committed:
Docs/formatting issues Mike mentioned

joemcgill committed:
Improve doc description for `wp_attachment_img_srcset_array()`

joemcgill commented on tests/phpunit/includes/factory.php:
```diff
@@ -94,6 +94,35 @@ class WP_UnitTest_Factory_For_Attachment extends WP_UnitTest_Factory_For_Post {
 		return wp_insert_attachment( $args, $file, $parent );
 	}
+
+	function create_upload_object( $file, $parent = 0 ) {
```
Adding this factory method is something wonderboymusic and I discussed after https://core.trac.wordpress.org/changeset/33807, which could replace redundant _make_attachment helper functions in tests/image/intermediate_size.php, tests/post/attachment.php, and tests/ajax/MediaElement.php. I still need to create a separate ticket for that.

joemcgill commented on tests/phpunit/includes/factory.php:
```diff
@@ -94,6 +94,35 @@ class WP_UnitTest_Factory_For_Attachment extends WP_UnitTest_Factory_For_Post {
 	function create_object( $file, $parent = 0, $args = array() ) {
 		return wp_insert_attachment( $args, $file, $parent );
 	}
+
```
Adding this factory method is something wonderboymusic and I discussed after https://core.trac.wordpress.org/changeset/33807, which could replace redundant _make_attachment helper functions in tests/image/intermediate_size.php, tests/post/attachment.php, and tests/ajax/MediaElement.php. I still need to create a separate ticket for that.

joemcgill replied:
Update: here's the ticket – https://core.trac.wordpress.org/ticket/34075

joemcgill committed:
Improve docblock descriptions for srcset arrays

joemcgill committed:
Build source urls using same logic as `image_get_intermediate_size()`

In `image_get_intermediate_size` we use `path_join()` to build the URL path by
concatonating the directory name from the original image file with the filename
of the intermediate size.

joemcgill committed:
Rename $img_edited to $img_is_edited

This should further improve readability along with the improved
inline docs in 59b2c09

joemcgill committed:
Revert "Escape srcset and sizes attributes on output"

We need to escape these attributes earlier, since we're adding the full
attribute, and not just the value within the attribute.

This reverts commit 07cf8053584baf125ee3cd6324ef1eec8e0acf61.

joemcgill committed:
Escape attributes in `get_image_tag()`

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+ *
+ * @param  int    $id   Image attachment ID.
+ * @param  string $size Optional. Name of image size. Default value: 'medium'.
+ * @return array|bool  An array of image urls and widths or false on failure.
```
https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#1-1-parameters-that-are-arrays

joemcgill replied:
Got it. Thanks.

getsource replied:
We still need to update these docs

joemcgill replied:
Looking into how to do this one correctly, and I'm not seeing any other functions that return a multi-dimensional array where the key is an integer, so I'm not sure what's best to do here. 

Right the returned array that looks like this:

```
  150 => 'http://test.dev/wp-content/uploads/2015/09/testimage-150x150.jpg 150w'
  300 => 'http://test.dev/wp-content/uploads/2015/09/testimage-300x300.jpg 300w'
  600 => 'http://test.dev/wp-content/uploads/2015/09/testimage.jpg 600w'
```

Should we rewrite the function so the output to use keyword => value pairs and provide a compatibility shim in our plugin, or do we reach out to Drew about how to best format this?

For example, we could reformat these arrays to read like this:

```
[0] {
  'url'       => 'http://test.dev/wp-content/uploads/2015/09/testimage-150x150.jpg'
  'width'     => 150
  'descriptor => 'w'
},
[1] {
  'url'       => 'http://test.dev/wp-content/uploads/2015/09/testimage-300x300.jpg'
  'width'     => 300
  'descriptor => 'w'
},
[2] {
  'url'       => 'http://test.dev/wp-content/uploads/2015/09/testimage.jpg'
  'width'     => 600
  'descriptor => 'w'
}

```

I probably prefer reformatting the array, since that will be cleaner going forward, and a compat shim should be fairly easy to write.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	list( $img_url, $img_width, $img_height ) = $img;
+
+	// Bail early If we have no width or if the width set to zero.
+	if ( ! $img_width > 0 ) {
```
Why does the height not matter here?  Is there a case for `srcset` of 0 ratio images?

joemcgill replied:
It would probably be safe to bail if there is no height, or if height was 0px as well, but technically speaking we only need to worry about widths because a) we use the width as the denominator when calculating aspect ratios, and b) we're using `w` descriptors in our `srcset` values.

getsource replied:
Mostly, I was thinking that it makes sense stop calculating/avoid a database call if we know that the srcset/sources returned are not going to be valid.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+		);
+
+		if ( strrpos( $img_meta['file'], '/' ) !== false ) {
+			$img_sizes['full']['file'] = substr( $img_meta['file'], strrpos( $img_meta['file'], '/' ) + 1 );
```
Can we use `basename()` or similar here?

joemcgill replied:
No problem. Will update.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+ * @param string $size Optional. Name of image size. Default value: 'thumbnail'.
+ * @return string|bool A 'srcset' value string or false.
+ */
+function wp_attachment_img_srcset( $id, $size = 'thumbnail' ) {
```
This default size is different from that of `wp_attachment_img_srcset_array()` -- is there reasoning for that?

joemcgill replied:
We used to use 'thumbnail', but 'medium' is a better default for these functions since 'thumbnail' will almost always return false because it's normally hard cropped. I just missed a few spots and will update this.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+function wp_attachment_img_srcset( $id, $size = 'thumbnail' ) {
+	$srcset_array = wp_attachment_img_srcset_array( $id, $size );
+
+	if ( empty( $srcset_array ) ) {
```
Is it better to return `false` or empty string here?  Let's check signatures of other similar functions that return attributes.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +810,117 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	}
+
+	// Only return sources if there are more than one.
+	if ( count( $sources ) <= 1 ) {
```
I'm feeing very mixed about this.  Is there a reason to not do this check in `wp_attachment_img_srcset( $id, $size = 'thumbnail' )`?  It seems like for consistency in signature, it might be nice to know there's one source if you're running `wp_attachment_img_srcset_array()` directly.

joemcgill replied:
I'm not opposed to that suggestion. We could also remove that stipulation completely and tell devs who are concerned about bloating their code with `srcset` attributes containing only one image to filter the output of `wp_attachment_img_srcset_array()` themselves.

joemcgill committed:
Use basename() to get the filename from img_meta['file']

joemcgill committed:
Return a single array value but not a single srcset value

Let `wp_attachment_img_srcset_array` return a single array value but do not return
a srcset value from `wp_attachment_img_srcset` if there is only one valid source
candidate.

joemcgill committed:
Update patch to match 2.5.2 release

joemcgill committed:
remove debugging cruft

joemcgill committed:
Update existing tests

Update a few tests that were failing due to shortcuts used for setting
up test post data because now our functions use data that was not found.
It may be worth going back through the failing tests and adding sanity
checks in the new `srcset` and `sizes` code.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	if ( $id && false === $size ) {
+		$size = array(
+			$width,
+			$height
```
Add trailing comma on last element of all arrays.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+ * @access private
+ * @since 4.4.0
+ */
+function _wp_resp_img( $image ) {
```
Why is this a separate function?

joemcgill replied:
This was originally a preg_match callback function. We could probably merge these now.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	* @param int     $id    Post ID of the original image.
+	* @param string  $size  Name of the image size being used.
+	*/
+	$args = apply_filters( 'wp_image_sizes_args', $args, $id, $size );
```
wp_resp_image_sizes_args?  Or what are the args really?  Maybe we can use that instead to avoid being generic?

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+ * }
+ * @return string|bool A valid source size value for use in a 'sizes' attribute or false.
+ */
+function wp_attachment_img_sizes( $id, $size = 'medium', $args = null ) {
```
If $args is more defined than "extra arguments," we can name it appropriately.  Is it $sizes?

joemcgill replied:
The args parameter is a set of arguments that can be passed to the the sizes function to customize the markup. When I originally wrote this function, I was thinking in terms of building an API that allows people to build an array of sizes, which would be converted into a properly formatted `sizes` value. Using a generic `$args` value seems to be a common pattern throughout WordPress and gave us some flexibility to change the input options as needed.

In practice, I've wondered if this whole function is over engineered and if we would be better served to drop the ability to build your own pattern, only build the default `sizes` attribute, and let devs filter the value however they like. If so, this function could be greatly simplified.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+	 * @param int          $id       Attachment ID for image.
+	 * @param array|string $size     Size of image, either array or string.
+	 */
+	return apply_filters( 'wp_attachment_img_srcset', $sources, $id, $size );
```
wp_attachment_image_srcset_array?  Or maybe responsive_image_array?  responsive_image_sources?  I need to check recently added filters and see if they all have wp_, but we can be descriptive on what it we're filtering either way.

I notice that it's possible to filter the array, but not the final values.  We should probably have filters for both.

joemcgill replied:
I believe this was a back compatibility issue when we separated the `tevkori_get_srcset_array` function from `tevkori_get_srcset`. We can rename this filter to match the name of the wrapping function.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+		 *
+		 * To avoid making a database call for each image, a single query
+		 * warms the object cache with the meta information for all images.
+		 **/
```
Multiline comments have no double *\* (docblocks do)
https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#5-2-multi-line-comments

getsource commented on src/wp-includes/default-filters.php:
```diff
@@ -135,6 +135,7 @@
 add_filter( 'the_content', 'wpautop'            );
 add_filter( 'the_content', 'shortcode_unautop'  );
 add_filter( 'the_content', 'prepend_attachment' );
+add_filter( 'the_content', 'wp_resp_img'        );
```
It looks like new filters without wp_ are fine, so let's see if we can find a name for these that are more descriptive of what you're filtering.

add_responsive_images?  responsive_image_markup?  add_responsive_images?  fill_resp_images?  (some other thing?)

Naming is hard.

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+ * @return string Converted content with 'srcset' and 'sizes' added to images.
+ */
+function wp_resp_img( $content ) {
+
```
Unless it makes it unreadable, let's remove leading line-breaks where they happen

getsource commented on src/wp-includes/media.php:
```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+
+	foreach( $matches[0] as $k => $image ) {
+		$match = array( $image, $matches[1][$k] );
+		$needle = $image;
```
Does this get used anywhere?

joemcgill committed:
Make output of `wp_attachment_img_srcset_array` more descriptive.

This changes the array pattern from a simple 'key' => 'value' pattern to
a multidimentional array containig the 'url', 'descriptor', and 'value' for
each image candidate. Updates tests.

joemcgill committed:
Remove debugging cruft.

joemcgill commented:
The patch has now been refreshed and should be up to date with Trunk.

joemcgill committed:
Rename functions to match  family.

joemcgill committed:
Filter output of `wp_get_attachment_image_srcset` and adjust filter names.

Filter the output of `wp_get_attachment_image_srcset()` with a filter of the same
name and change the name of the output filter on `wp_get_attachment_image_srcset_array()`,
which was `wp_get_attachment_image_srcset` to `wp_get_attachment_image_srcset_array`
to match the wrapping function's name and to avoid a conflict with the output
filter on `wp_get_attachment_image_srcset()`.

joemcgill committed:
Update inline docs for filters

joemcgill committed:
Clean up content filter functions.

Originally, we were using `preg_replace_callback` in `wp_resp_img()` to match
images in the content and pass the matches to a private callback function
named `_wp_resp_img()` so we could add `srcset` and `sizes`. Once we changed
to a `str_replace` method, the callback function seemed out of place.

This commit keeps the two functions separate, but renames the first
`wp_make_content_images_responsive()` which uses a modified version of
`get_media_embedded_in_content()` to match all the images in a content string
and pass each to the second function, now named `wp_img_add_srcset_and_sizes()`,
which can be used to add `srcset` and `sizes` to any image element.

`wp_make_content_images_responsive()` then uses `str_replace()` to replace the
original image markup with the transformed image markup returned by
`wp_img_add_srcset_and_sizes()`.

This adds `img` to the list of allowed media types that can be matched in
`get_media_embedded_in_content()`.

Test have been updated and some renamed to match our new functions.

joemcgill committed:
Fix undefined variable bug in `wp_make_content_images_responsive`

joemcgill commented on src/wp-includes/media.php:
```diff
@@ -815,6 +825,373 @@ function wp_get_attachment_image_url( $attachment_id, $size = 'thumbnail', $icon
+ * @return string Converted content with 'srcset' and 'sizes' added to images.
+ */
+function wp_make_content_images_responsive( $content ) {
+	$images = get_media_embedded_in_content( $content, 'img' );
```
Woke up this morning realizing that modifying `get_media_embedded_in_content()` to search for `img` elements might be heavy handed since all of the other patterns that get matched by that function are wrapping elements, whereas an `img` is a single element. That said, it does work, and there may be some utility in having a function in core that returns an array of image tags from a block of content. Would be curious to know your thoughts.

joemcgill replied:
After some playing around, we could do something like this:

```
if ( ! preg_match_all( '/<img [^>]+ wp-image-([0-9]+)[^>]+>/i', $content, $matches ) ) {
    return $content;
}

$images = $matches[0];
$attachment_ids = array_unique( $matches[1] );
```

...instead of altering `get_media_embedded_in_content()`, and get the added advantage of dropping the extra loop below to get all the image ids for our cache priming.

See PR #5

getsource replied:
I actually kind of like the idea of using `get_media_embedded_in_content()` for this.  _If_ the performance is the same, it saves us some copied regex, and would be a good API addition.

How does performance compare?

joemcgill replied:
The performance difference between the two options is negligible. I could leave #5 in unmerged and we can come back to it if Scott or anyone is unhappy with the change to `get_media_embedded_in_content()`.

getsource replied:
Sounds good

joemcgill committed:
Avoid using `image_downsize()` in `wp_get_attachment_image_sizes()`

getsource commented on src/wp-includes/default-filters.php:
```diff
@@ -135,6 +135,7 @@
 add_filter( 'the_content', 'wpautop'            );
 add_filter( 'the_content', 'shortcode_unautop'  );
 add_filter( 'the_content', 'prepend_attachment' );
+add_filter( 'the_content', 'wp_make_content_images_responsive' );
```
I really like the other changes to function names you did!  I think this one is probably too long, but naming shouldn't stop us from an initial patch.

joemcgill committed:
Merge branch 'master' into respimg

joemcgill committed:
Style and typo edits to inline comments in tests

joemcgill committed:
Make sure `wp_get_attachment_metadata` returns an array.

jaspermdegroot committed:
Correct version in inline doc

joemcgill committed:
Docs changes from Drew in 33641.2.patch

I've incorporated nearly all the changes suggested by Drew in
https://core.trac.wordpress.org/attachment/ticket/33641/33641.2.patch.

The only one I'm waiting on is the whitespace change in defautl-filters.php based
on the probability that the function name might get changed.

joemcgill committed:
Update outdated inline comment

joemcgill committed:
whitespacing in default-filters.php

joemcgill commented:
Merged in https://github.com/joemcgill/develop.wordpress/commit/fabd554bb2588f0ca52383128186adf9ba469329

SVN Changeset: [34855](https://core.trac.wordpress.org/changeset/34855)
